### PR TITLE
ovl_obj_lightblock OK

### PR DESCRIPTION
**Pull request by emilybrooks**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

### CODE REVIEW

File: src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c [hunk 1]
```diff
@@ -22,46 +27,178 @@ const ActorInit Obj_Lightblock_InitVars = {
...
+
+extern CollisionHeader D_06000B80;
+extern Gfx D_801AEF88[];
+extern Gfx D_06000178[];
+extern Gfx D_801AEFA0[];
+
+void func_80AF3910(ObjLightblock* this, GlobalContext* globalCtx) {
+    LightblockTypeVars* typeVars = &sLightblockTypeVars[LIGHTBLOCK_TYPE(&this->dyna.actor)];
+
+    Actor_Spawn(&globalCtx->actorCtx, globalCtx, 0x48, this->dyna.actor.world.pos.x, this->dyna.actor.world.pos.y,
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c:80]: Try to use the enum values defined in `include/z64actor.h`
```suggestion
    Actor_Spawn(&globalCtx->actorCtx, globalCtx, ACTOR_DEMO_EFFECT, this->dyna.actor.world.pos.x, this->dyna.actor.world.pos.y,
```

File: src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c [hunk 2]
```diff
@@ -22,46 +27,178 @@ const ActorInit Obj_Lightblock_InitVars = {
...
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Lightblock/func_80AF3C18.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Lightblock/func_80AF3C34.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Lightblock/ObjLightblock_Update.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Lightblock/ObjLightblock_Draw.s")
+extern ColliderCylinderInit sCylinderInit;
+extern InitChainEntry sInitChain[];
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c:70]: These are unnecessary now that the data is imported.

File: src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c [hunk 3]
```diff
@@ -22,46 +27,178 @@ const ActorInit Obj_Lightblock_InitVars = {
...
+
+void func_80AF3AC8(ObjLightblock* this) {
+    this->actionFunc = func_80AF3ADC;
+}
+
+void func_80AF3ADC(ObjLightblock* this, GlobalContext* globalCtx) {
+    if (this->collider.base.acFlags & AC_HIT) {
+        this->collider.base.acFlags &= ~AC_HIT;
+        // light arrows
+        if (this->collider.info.acHitInfo->toucher.dmgFlags & 1 << 13) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c:122]:
```suggestion
        if (this->collider.info.acHitInfo->toucher.dmgFlags & (1 << 13)) {
```

File: src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c [hunk 4]
```diff
@@ -22,46 +27,178 @@ const ActorInit Obj_Lightblock_InitVars = {
...
+    LightblockTypeVars* typeVars = &sLightblockTypeVars[LIGHTBLOCK_TYPE(&this->dyna.actor)];
+
+    Actor_Spawn(&globalCtx->actorCtx, globalCtx, 0x48, this->dyna.actor.world.pos.x, this->dyna.actor.world.pos.y,
+                this->dyna.actor.world.pos.z, 0, 0, 0, typeVars->params);
+}
+
+void ObjLightblock_Init(Actor* thisx, GlobalContext* globalCtx) {
+    ObjLightblock* this = THIS;
+    LightblockTypeVars* typeVars = &sLightblockTypeVars[LIGHTBLOCK_TYPE(&this->dyna.actor)];
+    s32 pad;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c:87]: Does this pad have to go here, or can it go higher? (Usually the globalCtx recast is uppermost, it seems.)

File: src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c [hunk 5]
```diff
@@ -22,46 +27,178 @@ const ActorInit Obj_Lightblock_InitVars = {
...
+void func_80AF3ADC(ObjLightblock* this, GlobalContext* globalCtx) {
+    if (this->collider.base.acFlags & AC_HIT) {
+        this->collider.base.acFlags &= ~AC_HIT;
+        // light arrows
+        if (this->collider.info.acHitInfo->toucher.dmgFlags & 1 << 13) {
+            this->collisionCounter = 8;
+        }
+        // light ray
+        else {
+            this->collisionCounter += 1;
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c:127]:
```suggestion
            this->collisionCounter++;
```

**Review by AngheloAlf — COMMENTED** [review]
Good work!
Just a few comments

File: src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c (hunk 3 above)
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c:122]: Hopefully we can get some damage flags to replace these soon.

**Review by EllipticEllipsis — COMMENTED** [review]
Nice job, just a couple of minor things. Could you make a description header like OoT uses? Never mind if you don't know exactly what this actor is, you don't have to fill in the description part.

**Review by EllipticEllipsis — APPROVED** [review]

**Review by AngheloAlf — APPROVED** [review]

File: src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.h [hunk 6]
```diff
@@ -3,15 +3,20 @@
...
 
 typedef struct ObjLightblock {
-    /* 0x0000 */ Actor actor;
-    /* 0x0144 */ char unk_144[0x64];
+    /* 0x0000 */ DynaPolyActor dyna;
+    /* 0x015C */ ColliderCylinder collider;
     /* 0x01A8 */ ObjLightblockActionFunc actionFunc;
-    /* 0x01AC */ char unk_1AC[0x4];
+    /* 0x01AC */ s8 timer;
+    /* 0x01AD */ u8 opacity;
```
Comment (rylieb) [inline src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.h:18]: Per something Fig mentioned to me in my En_Holl PR, "opacity" is typically referred to as "alpha" in a programming context, and specifically in the standard libraries used for this game. It's the "alpha" component of RGBA (not to be confused with "gamma", which adjusting can help you see in dark games better- that tripped me up for a while).
```suggestion
    /* 0x01AD */ u8 alpha;
```

File: src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c [hunk 7]
```diff
@@ -22,46 +33,175 @@ const ActorInit Obj_Lightblock_InitVars = {
...
+    if (Flags_GetSwitch(globalCtx, LIGHTBLOCK_DESTROYED(&this->dyna.actor))) {
+        Actor_MarkForDeath(&this->dyna.actor);
+    } else {
+        BgCheck3_LoadMesh(globalCtx, &this->dyna, &D_06000B80);
+        Collider_SetCylinder(globalCtx, &this->collider, &this->dyna.actor, &sCylinderInit);
+        Collider_UpdateCylinder(&this->dyna.actor, &this->collider);
+        this->collider.dim.radius = typeVars->radius;
+        this->collider.dim.height = typeVars->height;
+        this->collider.dim.yShift = typeVars->yShift;
+        this->opacity = 255;
```
Comment (rylieb) [inline src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c:105]:
```suggestion
        this->alpha = 255;
```

File: src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c [hunk 8]
```diff
@@ -22,46 +33,175 @@ const ActorInit Obj_Lightblock_InitVars = {
...
+    this->timer--;
+    if (this->timer <= 0) {
+        temp_a0 = this->dyna.actor.cutscene;
+        ActorCutscene_Stop(temp_a0);
+        Actor_MarkForDeath(&this->dyna.actor);
+    } else if (this->timer < 61) {
+        if (this->opacity >= 41) {
+            this->opacity -= 40;
+        } else {
+            this->opacity = 0;
```
Comment (rylieb) [inline src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c:176]:
```suggestion
        if (this->alpha >= 41) {
            this->alpha -= 40;
        } else {
            this->alpha = 0;
```

File: src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c [hunk 9]
```diff
@@ -22,46 +33,175 @@ const ActorInit Obj_Lightblock_InitVars = {
...
+
+void ObjLightblock_Draw(Actor* thisx, GlobalContext* globalCtx) {
+    ObjLightblock* this = THIS;
+
+    OPEN_DISPS(globalCtx->state.gfxCtx);
+    if (this->opacity < 255) {
+        func_8012C2DC(globalCtx->state.gfxCtx);
+        gSPSegment(POLY_XLU_DISP++, 0x08, D_801AEF88);
+        gSPMatrix(POLY_XLU_DISP++, Matrix_NewMtx(globalCtx->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+        gDPSetPrimColor(POLY_XLU_DISP++, 0, 0x80, 255, 255, 255, this->opacity);
```
Comment (rylieb) [inline src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c:197]:
```suggestion
    if (this->alpha < 255) {
        func_8012C2DC(globalCtx->state.gfxCtx);
        gSPSegment(POLY_XLU_DISP++, 0x08, D_801AEF88);
        gSPMatrix(POLY_XLU_DISP++, Matrix_NewMtx(globalCtx->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
        gDPSetPrimColor(POLY_XLU_DISP++, 0, 0x80, 255, 255, 255, this->alpha);
```

**Review by rylieb — CHANGES_REQUESTED** [review]
Good work on your first actor, happy to have you in the community!

This actor is giving me strong documentation urges. For instance right off the bat, I see it has 3 separate SetupAction functions that each hardcode a particular actionFunc. But given this is an undocumented PR, I'll keep my suggestions to a minimum.

The main thing is something I literally just ran into in my previous PR for EnHoll, which is that `opacity` should be called `alpha` to better align with the Alpha component of the RGBA color system used for graphics in this game.

**Review by rylieb — APPROVED** [review]

File: src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c [hunk 10]
```diff
@@ -22,46 +33,175 @@ const ActorInit Obj_Lightblock_InitVars = {
...
+void func_80AF3C34(ObjLightblock* this, GlobalContext* globalCtx) {
+    s8 temp_a0;
+
+    this->timer--;
+    if (this->timer <= 0) {
+        temp_a0 = this->dyna.actor.cutscene;
+        ActorCutscene_Stop(temp_a0);
+        Actor_MarkForDeath(&this->dyna.actor);
+    } else if (this->timer < 61) {
+        if (this->alpha >= 41) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c:173]:
```suggestion
        if (this->alpha > 40) {
```

File: src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c [hunk 11]
```diff
@@ -22,46 +33,175 @@ const ActorInit Obj_Lightblock_InitVars = {
...
+
+void func_80AF3C34(ObjLightblock* this, GlobalContext* globalCtx) {
+    s8 temp_a0;
+
+    this->timer--;
+    if (this->timer <= 0) {
+        temp_a0 = this->dyna.actor.cutscene;
+        ActorCutscene_Stop(temp_a0);
+        Actor_MarkForDeath(&this->dyna.actor);
+    } else if (this->timer < 61) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Lightblock/z_obj_lightblock.c:172]:
```suggestion
    } else if (this->timer <= 60) {
```

**Review by engineer124 — COMMENTED** [review]
Nice work! Just the tiniest nitpick, but basically good to go

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**